src/CoreWCF.Primitives/src/CoreWCF/Configuration/IConfigurationHolder.cs

```diff
@@ -0,0 +1,17 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using CoreWCF.Channels;
+
+namespace CoreWCF.Configuration
+{
+    public interface IConfigurationHolder
+    {
```

Comment:
I've been going back and forth on how to approach this. This configuration mechanism using IConfigureOptions kind of mirrors the UseServiceModel apis which takes a delegate which is passed a IServiceBuilder instance where you define the service. The difference is the convention with IConfigureOptions is to have the options type be a concrete class implementation. So I took a look at how the KestrelServerOptions is used. In `NetTcpFramingOptionsSetup.Configure` we configure an instance of KestrelServerOptions. We've previously populated an endpoint list on our own internal configuratopm class NetTcpFramingOptions via calls to UseNetTcp. Basically if you call UseNetTcp with two different ports, it will have to endpoints listed in NetTcpFramingOptions.Endpoints. We iterate over the endpoints and call KestrelServerOptions.Listen for each one in turn. The code looks like this:
```
            foreach (IPEndPoint endpoint in _options.EndPoints)
            {
                options.Listen(endpoint, builder =>
                {
                    builder.UseConnectionHandler<NetMessageFramingConnectionHandler>();
                    NetMessageFramingConnectionHandler handler = builder.ApplicationServices.GetRequiredService<NetMessageFramingConnectionHandler>();
                    // Save the ListenOptions to be able to get final port number for adding BaseAddresses later
                    ListenOptions.Add(builder);
                });
            }
```

We should be going for a similar model. The method ConfigurationManagerServiceModelOptions.Configure should have some way to parse the configuration into some kind of object model. The ConfigurationHolder class looks like a good candidate. This class should be internal though. It could either be done in Configure with a method call, or via something that is injected. Then the this object model is iterated over and ServiceModelOptions populated with a more typical set of objects which reflect closely to what IServiceBuilder does. At service startup, CoreWCF should get the ServiceModelOptions and iterate over it's data to set things up. So ConfigurationManagerServiceModelOptions.Configure would look something like this:
```
public void Configure(ServiceModelOptions options)
{
    var configHolder = ParseConfig();
    foreach (var endpointName in configHolder.Endpoints.Keys) // expose ConfigurationHolder._endpoints as property
    {
        var configEndpoint = configHolder.GetXmlConfigEndpoint(endpointName);
        options.ConfigureService(configEndpoint.Service, serviceConfig =>
        {
            serviceConfig.ConfigureServiceEndpoint(endpoint.Service, endpoint.Contract, endpoint.Binding, endpoint.Address, null);
        });
    }
}
```
Having ServiceModelOptions and another class ServiceConfigurationBuilder be public types which the ServiceBuilder can pull from DI and use to build the service allows a future capability to add other config options such as JSON. The classes would look like this and live in primitives.
```
public class ServiceModelOptions
{
    private Dictionary<Type, ServiceConfigurationBuilder> _config = new Dictionary<Type, ServiceConfigurationBuilder>();
    public void ConfigureService(Type serviceType, Action<ServiceConfigurationBuilder> configure)
    {
        ServiceConfigurationBuilder configBuilder;
        if (!_config.TryGetValue(serviceType, out configBuilder))
        {
            configBuilder = new ServiceConfigurationBuilder(serviceType);
            _config[serviceType] = configBuilder;
        }
        configBuilder.AddConfigureDelegate(configure);
    }
    internal void ConfigureServiceBuilder(IServiceBuilder serviceBuilder)
    {
        foreach (var serviceConfigBuilder in _config.Values)
        {
            serviceConfigBuilder.ConfigureServiceBuilder(serviceBuilder);
        }
    }
}

public class ServiceConfigurationBuilder
{
    private List<Action<ServiceConfigurationBuilder>> _configDelegates = new List<Action<ServiceConfigurationBuilder>>();
    private Dictionary<Type, List<(Type implementedContract, Binding binding, Uri address, Uri listenUri)>> _endpoints;
    internal Type ServiceType { get; }
    public ServiceConfigurationBuilder(Type serviceType)
    {
        ServiceType = serviceType;
    }
    public void AddConfigureDelegate(Action<ServiceConfigurationBuilder> configDelegate)
    {
        _configDelegates.Add(configDelegate);
    }
    public void ConfigureServiceEndpoint(Type service, Type implementedContract, Binding binding, Uri address, Uri listenUri)
    {
        List<(Type implementedContract, Binding binding, Uri address, Uri listenUri)> serviceEndpoints;
        if (!_endpoints.TryGetValue(service, out serviceEndpoints))
        {
            serviceEndpoints = new List<(Type implementedContract, Binding binding, Uri address, Uri listenUri)>();
            _endpoints[service] = serviceEndpoints;
        }
        serviceEndpoints.Add((implementedContract, binding, address, listenUri));
    }
    internal void ConfigureServiceBuilder(IServiceBuilder serviceBuilder)
    {
        _endpoints = new Dictionary<Type, List<(Type implementedContract, Binding binding, Uri address, Uri listenUri)>>();
        foreach (var configDelegate in _configDelegates)
        {
            configDelegate(this);
        }
        foreach (var service in _endpoints)
        {
            serviceBuilder.AddService(service.Key);
            serviceBuilder.AddServiceEndpoint(service.Key, service.Value.implementedContract, service.Value.binding, service.Value.address, service.Value.listenUri);
        }
        _endpoints = null;
    }
}
```
Then in ServiceBuilder where is creates the service, right at the top it grabs from DI ServiceModelOptions and calls `ServiceModelOptions.ConfigureServiceBuilder(this)`.

Reply:
Btw, I wrote this code in the GitHub comment box so there might be typos or some errors in there.

Reply:
@mconnew @Ximik87 should we expose another method inside ServiceConfigurationBuilder as ConfigureBehavior(), which would be called from ConfigurationManagerServiceModelOptions.Configure (if behaviors present) and set up the ServiceBehaviors and EndPointBehaviors.. this would enable the TransportWithMessageCred etc... let me know what you both think...

Reply:
yes, I have little experience with IConfigureOptions and I did not fully understand how it works correctly.
@mconnew , I did as you described, and it works except ServiceModelOptions.ConfigureServiceBuilder (this)
because - AddService in Configure (app, env) calls ServiceModelOptions.ConfigureServiceBuilder (this),
this calls ServiceConfigurationBuilder.onfigureServiceBuilder () and this calls AddService again and so recursively ...
I made an extension method for IServiceBuilder that explicitly calls ServiceModelOptions.ConfigureServiceBuilder and it seems to work, but check it out.

Reply:
@birojnayak, Yes, as I understand it Matt - all ServiceBehaviors and EndPointBehaviors settings will be read in ConfigurationManagerServiceModelOptions.Configure
but we will use them in the ServiceConfigurationBuilder, make a separate method for this - I don't understand yet, I have to start trying coding